kernel/src/kernel.rs

```diff
@@ -41,10 +46,43 @@ use crate::utilities::cells::NumericCellExt;
 /// is less than this threshold.
 pub(crate) const MIN_QUANTA_THRESHOLD_US: u32 = 500;
 
+#[derive(Copy, Clone)]
+pub struct StaticSlice<T> {
+    len: usize,
+    base: *mut T,
+}
+
+impl<T> StaticSlice<T> {
+    pub fn new<const N: usize>(slice: &'static mut [T; N]) -> Self {
+	StaticSlice {
+	    len: N,
+	    base: slice.as_mut_ptr(),
+	}
+    }
+
+    pub(crate) fn len(&self) -> usize {
+        self.len
+    }
+}
+
+impl<T> Deref for StaticSlice<T> {
+    type Target = [T];
+
+    fn deref(&self) -> &Self::Target {
+        unsafe { core::slice::from_raw_parts(self.base, self.len) }
+    }
+}
+
+impl<T> DerefMut for StaticSlice<T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        unsafe { core::slice::from_raw_parts_mut(self.base, self.len) }
+    }
+}
```

Comment:
`StaticSlice` is unsound. Because `StaticSlice` is `Copy`, you create a `StaticSlice` pointing to a buffer, duplicate it, then call `deref_mut` on both copies to get two `&mut` references to the same buffer.

Demonstration: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=21831ccf86ae6ebcc70ef9d9263a9fb4

Reply:
This was an oversight when I was just trying to get the uses of it elsewhere in kernel to work. I've fixed this, it is no longer `Copy` or `Clone` and, you're right, absolutely should not be.